Add OLM bundle

alebedev87 opened the pull request:
This PR adds `manifests/` and `metadata/` directories which contain the files to be present in OLM bundle for ExternalDNS Operator.
The PR adds some README instructions on how to install the bundle on the cluster. These instructions were tested manually:
```bash
$ oc -n openshift-marketplace get catalogsource external-dns-operator
NAME                    DISPLAY   TYPE   PUBLISHER   AGE
external-dns-operator             grpc               7m47s

$ oc -n openshift-marketplace get pods | grep external-dns-operator
external-dns-operator-jrkzc                                       1/1     Running     0          7m55s

$ oc -n external-dns-operator get csv,sub,pods
NAME                                                                      DISPLAY                VERSION   REPLACES   PHASE
clusterserviceversion.operators.coreos.com/external-dns-operator.v0.0.1   ExternalDNS Operator   0.0.1                Succeeded

NAME                                                      PACKAGE                 SOURCE                  CHANNEL
subscription.operators.coreos.com/external-dns-operator   external-dns-operator   external-dns-operator   alpha

NAME                                         READY   STATUS    RESTARTS   AGE
pod/external-dns-operator-796bcbfbd6-b4p7j   2/2     Running   0          6m47s
```

The contents of `manifests/` directory is the same as the rendered version of Kustomize files in `config/default/`.

sgreene570 commented on manifests/external-dns-operator_v1_serviceaccount.yaml:
```diff
@@ -0,0 +1,5 @@
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  creationTimestamp: null
```
is the null creationTimestamp necessary in the manifest files?

alebedev87 replied:
No, I don't quite know why but most of the Kube/OpenShift utilities add `null` creation timestamp when they scaffold/export things. That was added by operator-sdk when I used it to scaffold the OLM bundle.
I can remove it.

sgreene570 replied:
interesting. sounds good.

sgreene570 commented on api/v1alpha1/groupversion_info.go:
```diff
@@ -26,7 +26,7 @@ import (
 var (
 	// GroupVersion is group version used to register these objects
-	GroupVersion = schema.GroupVersion{Group: "operator.openshift.io", Version: "v1alpha1"}
+	GroupVersion = schema.GroupVersion{Group: "externaldns.olm.openshift.io", Version: "v1alpha1"}
```
The API group changes in this PR will be taken care of via https://github.com/openshift/external-dns-operator/pull/12, right? (I dont see any commits from that PR explicitly cherry-picked into this PR so just wanted to double check)

alebedev87 replied:
Yeah, you are right. I put a note in the description that the API changes are here just to match the future names. Taken that I renamed the files in `manifests/` to `externaldns...` I thought that it would be good to try the installation of the operator with the relevant API group.
This PR will wait for the merge of #12 anyway so that I will be able to rebase.

sgreene570 replied:
Ack, thanks for confirming.

alebedev87 commented:
/assign @sgreene570 @Miciah

sgreene570 commented on config/samples/operator_v1alpha1_externaldns.yaml:
```diff
@@ -3,3 +3,10 @@ kind: ExternalDNS
+      credentials: 
+        name: aws-access-key
+  source:
+    type: Service
```
nice defaults!

sgreene570 commented on manifests/external-dns-operator_clusterserviceversion.yaml:
```diff
@@ -0,0 +1,121 @@
+  description: ExternalDNS Operator deploys and manages ExternalDNS, which dynamically manages external DNS records in specific DNS Providers for specific Kubernetes resources.
+  displayName: ExternalDNS Operator
+  icon:
+  - base64data: 'iVBORw0KGgoAAAANSUhEUgAAAG4AAAB1CAMAAACYlCSRAAAABGdBTUEAALGPC/xhBQAAAAFzUkdCAK7OHOkAAAMAUExURQAAAChbgRdOdyhbgStdgx5TeyhbgShbgShbgShbgSJXfiBVfCVZgChbgShbgShbgShbgShbgShbgShbgShbgShbgShbgShbgShbgShbgShbgShbgShbgShbgShbgShbgShbgShbgShbgShbgShbgShbgShbgShbgf/////+/itdg/z9/vz9/Shbgf3+/ilcghpQeSpdgrrh5P7+/+pwb7ng4xhPeFV9nJavwulxcP7///v8/Nbb3RxSevb5+vr7++Xo6urs7cbN0bnBxqeyt63T17a/w6Ots5 ...
```
This is https://github.com/kubernetes-sigs/external-dns/blob/master/img/external-dns.png for those following along

alebedev87 replied:
Yeah, I tried with this one first but converting it to base64 gave me a string so long that my vim broke up :)
I'll think of how to get a smaller version of this PNG.

sgreene570 commented:
overall looks good to me

candita commented on README.md:
````diff
@@ -16,3 +16,39 @@ This Operator is in the early stages of implementation. For the time being, plea
+
+2. Build and push the image index for operator-registry:
+   ```sh
+   # to get opm: https://github.com/operator-framework/operator-registry
````
"to get opm" -> "download opm if needed from"

alebedev87 replied:
changed

alebedev87 commented:
Rebased from main (changes in CRD and RBAC).

miheer commented on README.md:
```diff
@@ -26,3 +26,41 @@ This Operator is in the early stages of implementation. For the time being, plea
         *Note*: other provider options can be found in `api/v1alpha1/externaldns_types.go`, e.g. the `ExternalDNSAWSProviderOptions` structure for AWS.
```
For now we have just tested on AWS and GCP right ?

alebedev87 replied:
I tested only on AWS for the moment.

miheer commented:
@alebedev87   aren't we adding Subscription object ? https://docs.openshift.com/container-platform/4.5/operators/understanding/olm/olm-understanding-olm.html#olm-subscription_olm-understanding-olm

alebedev87 commented:
@miheer : I mentioned the install via the OpenShift console which effectively creates a `Subscription` resource.

Changed the README to apply a subscription manifest + a note of how to do it in the console.

Also, added some make targets to simplify README.

miheer commented:
/lgtm
/hold for others to review

Miciah commented on Makefile:
```diff
@@ -103,6 +106,30 @@ deploy: manifests kustomize ## Deploy controller to the K8s cluster specified in
+	cp -f config/rbac/service_account.yaml manifests/external-dns-operator_v1_serviceaccount.yaml
+	cp -f config/rbac/auth_proxy_role.yaml manifests/external-dns-operator-auth-proxy_rbac.authorization.k8s.io_v1_clusterrole.yaml
+	cp -f config/rbac/auth_proxy_role_binding.yaml manifests/external-dns-operator-auth-proxy_rbac.authorization.k8s.io_v1_clusterrolebinding.yaml
+	cp -f config/rbac/auth_proxy_service.yaml manifests/external-dns-operator-auth-proxy_v1_service.yaml
```
I'm a little confused how all the pieces fit together, why `manifests/` has a file that isn't copied from `config/`, and why `config/` has files that don't appear to be incorporated into the OLM bundle anywhere.  Do we need anything from `config/manager/`?  Should `manifests/external-dns-operator_clusterserviceversion.yaml` be copied in from somewhere?  Should `config/rbac/externaldns_editor_role.yaml` and `config/rbac/externaldns_viewer_role.yaml` be copied in as well?

alebedev87 replied:
Normally `operator-sdk` is supposed do all the job:
- Create `ClusterServiceVersion` manifest with `operator-sdk generate kustomize manifests` command
- Get the output of your manifests from `config/default/` and generate a bundle (manifests + ClusterServiceVersion + metadata + Dockerfile) out of it with `operator-sdk generate bundle` command

But I decided to not add `operator-sdk` to the repository for these reasons:
- it's a kubebuilder project, so to not add confusion about why operator-sdk is used in one place but not in others
- operator-sdk generation of ClusterServiceVersion is not optimal: it needs some parameters from you but not all you want is parameterized. Some things are defaulted and don't work in your case some has be added/changed in order for `opm` to pass the checks.
- operator-sdk generation of some other things is not optimal for this project, e.g. unused scorecard stuff is added.

So, overall I found `operator-sdk generate bundle` is good for the scaffolding but not for the regular regeneration. That's why I put this small make rule to simplify a little the life of creation of OLM bundle. As you rightfully noticed CSV is not touched by this command. I let it to be modified manually as it's complex therefore a bit difficult to be automated.

The contents of `config/manager` is in CSV for the install strategy part. I moved all the other resources to the dedicated files except for the deployment as the install strategy has to contain something.

Will update the comment with more details.

Miciah reviewed (commented):
Small suggestion for the Makefile's `opm` target and a couple questions because I still don't completely understand OLM.

alebedev87 commented on README.md:
````diff
@@ -26,3 +26,56 @@ This Operator is in the early stages of implementation. For the time being, plea
+   $ make index-image-push
+   ```
+
+3. Create the catalogsource (registry secret may need to be linked to the external-dns-operator's pod created in `openshift-marketplace`):
````
It's some additional pod which is created when a catalog source is added to `openshift-marketplace`. It has the same prefix as the catalog source, a single container with the index image and seems to serve the data from the index (corresponding service is created too).

alebedev87 commented on Makefile:
```diff
@@ -121,3 +148,8 @@ GOBIN=$(PROJECT_DIR)/bin go get $(2) ;\
+
+OPM = $(shell pwd)/bin/opm
+opm: ## Download opm locally if necessary.
+	curl -sSfL https://github.com/operator-framework/operator-registry/releases/download/$(OPM_VERSION)/linux-amd64-opm -o $(PROJECT_DIR)/bin/opm
```
Done

Miciah commented:
/lgtm

miheer commented:
/unhold
